src/js/utils/animations.js

```diff
@@ -0,0 +1,150 @@
+const ZOOM_SIZES = {
+  xsmall: 0.001,
+  small: 0.01,
+  medium: 0.05,
+  large: 0.1,
+  xlarge: 0.5,
+};
+
+export const animationBounds = (type, size = 'medium') => {
+  if (type === 'draw') {
+    return ['', `stroke-dashoffset: 0`];
+  }
+  if (type === 'fadeIn') {
+    return ['opacity: 0;', 'opacity: 1;'];
+  }
+  if (type === 'fadeOut') {
+    return ['opacity: 1;', 'opacity: 0;'];
+  }
+  if (type === 'jiggle') {
+    const deg = JIGGLE_SIZES[size];
+    return [`transform: rotate(-${deg}deg);`, `transform: rotate(${deg}deg);`];
+  }
+  if (type === 'pulse') {
+    return ['transform: scale(1);', `transform: scale(${PULSE_SIZES[size]})`];
+  }
+  if (type === 'rotateRight') {
+    return [`transform: rotate(0deg);`, `transform: rotate(359deg);`];
+  }
+  if (type === 'rotateLeft') {
+    return [`transform: rotate(0deg);`, `transform: rotate(-359deg);`];
+  }
+  if (type === 'flipIn') {
+    return ['transform: rotateY(90deg);', 'transform: rotateY(0);'];
+  }
+  if (type === 'flipOut') {
+    return ['transform: rotateY(0);', 'transform: rotateY(90deg);'];
+  }
+  if (type === 'slideDown') {
+    return [
+      `transform: translateY(-${SLIDE_SIZES[size]}%);`,
+      'transform: none;',
+    ];
+  }
+  if (type === 'slideLeft') {
+    return [
+      `transform: translateX(${SLIDE_SIZES[size]}%);`,
+      'transform: none;',
+    ];
+  }
+  if (type === 'slideRight') {
+    return [
+      `transform: translateX(-${SLIDE_SIZES[size]}%);`,
+      'transform: none;',
+    ];
+  }
+  if (type === 'slideUp') {
+    return [
+      `transform: translateY(${SLIDE_SIZES[size]}%);`,
+      'transform: none;',
+    ];
+  }
+  if (type === 'zoomIn') {
+    return [`transform: scale(${1 - ZOOM_SIZES[size]});`, 'transform: none;'];
+  }
+  if (type === 'zoomOut') {
+    return [`transform: scale(${1 + ZOOM_SIZES[size]});`, 'transform: none;'];
+  }
+  return [];
+};
+
+export const normalizeTiming = (time, defaultTiming) =>
+  time ? `${time / 1000.0}s` : defaultTiming;
+
+// Animation Ending
+export const animationEnding = type => {
+  if (type === 'draw') {
+    return 'linear forwards';
+  }
+  if (type === 'jiggle') {
+    return 'alternate infinite';
+  }
+  if (type === 'pulse') {
+    return 'alternate infinite';
+  }
+  if (type === 'rotateRight' || type === 'rotateLeft') {
+    return 'infinite linear';
+  }
+  return 'forwards';
+};
+
+export const animationObjectStyle = (animation, theme) => {
+  const bounds = animationBounds(animation.type, animation.size);
+
+  if (bounds) {
+    const animationTransition = css`
+      from {
+        ${bounds[0]};
+      }
+      to {
+        ${bounds[1]};
+      }
+    `;
+
+    const defaultDuration = () => {
+      if (theme.diagram.animation !== undefined) {
+        return normalizeTiming(
+          theme.diagram.animation[animation.type]
+            ? theme.diagram.animation[animation.type].duration
+            : undefined,
+          theme.diagram.animation.duration,
+        );
+      }
```

Comment:
I don't think that the `diagram` specific code should sit on a generic utils file. The diagram logic should be performed on the diagram itself and sent to utils for the calculation.

Regardless, I don't think we should have a separate duration for each type of animation, but instead one duration for all types of animation to keep the API simple. Is Box allowing different duration for each type of animation?